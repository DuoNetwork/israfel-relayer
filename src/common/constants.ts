export const DB_ISRAFEL = 'israfel';
export const DB_LIVE_ORDERS = 'liveOrders';
export const DB_RAW_ORDERS = 'rawOrders';
export const DB_USER_ORDERS = 'userOrders';
export const DB_SEQUENCE = 'sequence';
export const DB_STATUS = 'status';
export const DB_LIVE = 'live';
export const DB_DEV = 'dev';
export const DB_PAIR = 'pair';
export const DB_INITIAL_SEQ = 'initialSequence';
export const DB_CURRENT_SEQ = 'currentSequence';
export const DB_ORDER_HASH = 'orderHash';
export const DB_PRICE = 'price';
export const DB_ACCOUNT = 'account';
export const DB_ACCOUNT_YM = 'accountYearMonth';
export const DB_PAIR_OH_SEQ_STATUS = 'pairOrderHashSequenceStatus';
export const DB_TYPE = 'type';
export const DB_CONFIRMED = 'confirmed';
export const DB_MATCHING = 'matching';
export const DB_USER = 'user';
export const DB_ORDER_PROCESSOR = 'orderProcessor';
export const DB_ORDER_MATCHER = 'orderMatcher';
export const DB_ORDER_WATCHER = 'orderWatcher';
export const DB_RELAYER = 'relayer';
export const DB_URL = 'url';
export const DB_SIDE = 'side';
export const DB_ASK = 'ask';
export const DB_BID = 'bid';
export const DB_BALANCE = 'balance';
export const DB_AMOUNT = 'amount';
export const DB_FILL = 'fill';
export const DB_PFILL = 'pFill';
export const DB_CREATED_AT = 'createdAt';
export const DB_UPDATED_AT = 'updatedAt';
export const DB_UPDATED_BY = 'updatedBy';
export const DB_PROCESSED = 'processed';
export const DB_TOKENS = 'tokens';
export const DB_ADDRESS = 'address';
export const DB_CODE = 'code';
export const DB_EXP = 'expiry';
export const DB_DENOMINATION = 'denomination';
export const DB_PRECISION = 'precision';
export const DB_FEE = 'fee';
export const DB_FEE_ASSET = 'feeAsset';

export const DB_ADD = 'add';
export const DB_UPDATE = 'update';
export const DB_TERMINATE = 'terminate';

export const DB_PROCESS = 'process';
export const DB_HOSTNAME = 'hostname';
export const DB_COUNT = 'count';

export const DB_0X_SENDER_ADDR = 'senderAddress';
export const DB_0X_MAKER_FEE = 'makerFee';
export const DB_0X_TAKER_FEE = 'takerFee';
export const DB_0X_MAKER_ADDR = 'makerAddress';
export const DB_0X_TAKER_ADDR = 'takerAddress';
export const DB_0X_MAKER_ASSET_AMT = 'makerAssetAmount';
export const DB_0X_TAKER_ASSET_AMT = 'takerAssetAmount';
export const DB_0X_MAKER_ASSET_DATA = 'makerAssetData';
export const DB_0X_TAKER_ASSET_DATA = 'takerAssetData';
export const DB_0X_SALT = 'salt';
export const DB_0X_EXCHANGE_ADDR = 'exchangeAddress';
export const DB_0X_FEE_RECIPIENT_ADDR = 'feeRecipientAddress';
export const DB_0X_EXPIRATION_TIME_SECONDS = 'expirationTimeSeconds';
export const DB_0X_SIGNATURE = 'signature';

export const DB_ORDERS = 'orders';
export const DB_ORDER_BOOKS = 'orderBooks';
export const DB_QUEUE = 'queue';
export const DB_CACHE = 'cache';
export const DB_PUBSUB = 'pubsub';
export const DB_SNAPSHOT = 'snapshot';

export const AWS_DYNAMO_API_VERSION = '2012-10-08';

export const TAKER_ETH_DEPOSIT = 10; // for development only
export const ONE_SECOND_MS = 1000;
export const ONE_MINUTE_MS = ONE_SECOND_MS * 60;
export const TEN_MINUTES_MS = ONE_MINUTE_MS * 10;
export const RELAYER_ADDR_KOVAN = '0x003519A4aB2C35c59Cb31d9194A45DD3F9Bf9e32';
export const RELAYER_ADDR_MAIN = '0x003519A4aB2C35c59Cb31d9194A45DD3F9Bf9e32';
export const DUMMY_ADDR = '0x0000000000000000000000000000000000000000';
export const PROVIDER_INFURA_KOVAN = 'https://kovan.infura.io';
export const PROVIDER_INFURA_MAIN = 'https://mainnet.infura.io';
export const NETWORK_ID_MAIN = 1;
export const NETWORK_ID_KOVAN = 42;
export const PROVIDER_LOCAL = 'http://localhost:8545';
export const NETWORK_ID_LOCAL = 50;
export const PENDING_HOURS = 24;

export const WS_HISTORY = 'history';
export const WS_INFO = 'info';
export const WS_SUB = 'subscribe';
export const WS_UNSUB = 'unsubscribe';
export const WS_CHANNEL = 'channel';
export const WS_METHOD = 'method';
export const WS_INVALID_REQ = 'invalid request';
export const WS_INVALID_ORDER = 'invalid order';
export const WS_ERROR = 'error';
export const WS_OK = 'ok';
export const WS_SERVICE_NA = 'service not available';

export const WS_CHANNEL_ORDERBOOK = 'orderbook';
export const WS_TYPE_ORDERBOOK = 'subscribe';
export const WS_CHANNEL_ORDER = 'order';
export const WS_TYPE_ORDER_UPDATE = 'update';
export const WS_TYPE_ORDER_ADD = 'add';
export const WS_TYPE_ORDER_CANCEL = 'cancel';

export const TOKEN_WETH = 'WETH';

export const PRUNE_INTERVAL = 3600000;
export const ORDER_PRUNE = 'pruneOrder';
export const SET_ALLOWANCE = 'setAllowance';
export const START_RELAYER = 'startRelayer';

export const LOG_INFO = 'INFO';
export const LOG_DEBUG = 'DEBUG';
export const LOG_ERROR = 'ERROR';
export const LOG_RANKING: { [level: string]: number } = {
	[LOG_ERROR]: 0,
	[LOG_INFO]: 1,
	[LOG_DEBUG]: 2
};

export const PRICE_PRECISION = 8;

export const BASE_DERIVATION_PATH = `44'/60'/0'/0`;
