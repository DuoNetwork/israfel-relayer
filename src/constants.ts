export const DB_ORDERS = 'orders';
export const DB_MAKER_ADDR = 'makerAddr';
export const DB_TIMESTAMP = 'timestamp';
export const DB_UPDATED_AT = 'updatedAt';
export const PROVIDER_INFURA_KOVAN = 'https://kovan.infura.io';
export const NETWORK_ID_KOVAN = 42;
export const PROVIDER_LOCAL = 'http://localhost:8545';
export const NETWORK_ID_LOCAL = 50;
export const RELAYER_HTTP_URL = 'http://localhost:3000/v0';
export const RELAYER_WS_URL = 'ws://localhost:8080';
export const MNEMONIC = 'concert load couple harbor equip island argue ramp clarify fence smart topic'
export const BASE_DERIVATION_PATH =  `44'/60'/0'/0`;
export const PENDING_HOURS = 24;

export const WS_CHANNEL_ORDERBOOK = 'orerbook';
export const WS_TYPE_ORDERBOOK = 'subscribe';
export const WS_CHANNEL_ORDER = 'order';
export const WS_TYPE_ORDER_UPDATE = 'update';
export const WS_TYPE_ORDER_ADD = 'add';
export const WS_TYPE_ORDER_CANCEL = 'cancel';

export const ONE_SECOND_MS = 1000;
export const ONE_MINUTE_MS = ONE_SECOND_MS * 60;
export const TEN_MINUTES_MS = ONE_MINUTE_MS * 10;
export const NULL_ADDRESS = '0x0000000000000000000000000000000000000000';
