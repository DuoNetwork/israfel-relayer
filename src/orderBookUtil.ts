import moment from 'moment';
import assetsUtil from './common/assetsUtil';
import * as CST from './constants';
import dynamoUtil from './dynamoUtil';
import redisUtil from './redisUtil';
import { IDuoOrder, ILiveOrders, IOption, IOrderBookSnapshot, IOrderBookUpdateWS } from './types';
// import util from './util';

class OrderBookUtil {
	public orderBook: { [key: string]: IOrderBookSnapshot } = {};

	public async init(tool: string, option: IOption) {
		const config = require('./keys/' + (option.live ? 'live' : 'dev') + '/dynamo.json');
		dynamoUtil.init(config, option.live, tool);
	}

	public async calculateOrderBookSnapshot() {
		for (const marketId of CST.TRADING_PAIRS) {
			const liveOrders: ILiveOrders[] = await dynamoUtil.getLiveOrders(marketId);
			this.orderBook[marketId] = this.aggrOrderBook(liveOrders);
			console.log('### current orerbook ', this.orderBook[marketId]);
		}
	}

	public sortByPrice(liveOrders: ILiveOrders[], isDescending: boolean): ILiveOrders[] {
		liveOrders.sort((a, b) => {
			return isDescending ? b.price - a.price : a.price - b.price;
		});
		console.log('### after sortByPrice: ', liveOrders);
		return liveOrders;
	}

	public aggrOrderBook(rawLiveOrders: ILiveOrders[]): IOrderBookSnapshot {
		// const rawOrderBook = this.getOrderBook(rawOrders, marketId);

		return {
			timestamp: moment.utc().valueOf(),
			bids: this.aggrByPrice(
				this.sortByPrice(
					rawLiveOrders.filter(order => order[CST.DB_SIDE] === CST.DB_BUY),
					true
				).map(bid => this.parseOrderBookUpdate(bid))
			),
			asks: this.aggrByPrice(
				this.sortByPrice(
					rawLiveOrders.filter(order => order[CST.DB_SIDE] === CST.DB_SELL),
					false
				).map(ask => this.parseOrderBookUpdate(ask))
			)
		};
	}

	public aggrByPrice(orderInfo: IOrderBookUpdateWS[]) {
		return orderInfo.reduce((past: IOrderBookUpdateWS[], current) => {
			const same = past.find(r => r && r.price === current.price);
			if (same) same.amount = (Number(same.amount) + Number(current.amount)).toString();
			else past.push(current);
			return past;
		}, []);
	}

	public parseOrderBookUpdate(order: ILiveOrders): IOrderBookUpdateWS {
		return {
			amount: order.amount.toString(),
			price: order.price.toString()
		};
	}

	public applyChangeOrderBook(
		marketId: string,
		timestamp: number,
		bidChanges: IOrderBookUpdateWS[],
		askChanges: IOrderBookUpdateWS[]
	) {
		const newBids = [...this.orderBook[marketId].bids, ...bidChanges].sort((a, b) => {
			return a.price > b.price ? 1 : b.price > a.price ? -1 : 0;
		});
		const newAsks = [...this.orderBook[marketId].asks, ...askChanges].sort((a, b) => {
			return a.price > b.price ? -1 : b.price > a.price ? 1 : 0;
		});
		this.orderBook[marketId] = {
			timestamp: timestamp,
			bids: this.aggrByPrice(newBids),
			asks: this.aggrByPrice(newAsks)
		};
	}

	public getOrderBook(orders: IDuoOrder[], marketId: string): IDuoOrder[][] {
		const baseToken = marketId.split('-')[0];
		const bidOrders = orders.filter(order => {
			const takerTokenName = order.takerAssetData
				? assetsUtil.assetDataToTokenName(order.takerAssetData)
				: null;
			return takerTokenName === baseToken;
		});

		const askOrders = orders.filter(order => {
			const makerTokenName = order.makerAssetData
				? assetsUtil.assetDataToTokenName(order.makerAssetData)
				: null;
			return makerTokenName === baseToken;
		});

		return [bidOrders, askOrders];
	}

	public scheduleSumamrizer() {
		setInterval(async () => {
			await this.calculateOrderBookSnapshot();
			for (const marketId in this.orderBook)
				redisUtil.publish(
					`${CST.ORDERBOOK_SNAPSHOT}|${marketId}`,
					JSON.stringify(this.orderBook[marketId])
				);
		}, 30000);
	}
}
const orderbookUtil = new OrderBookUtil();
export default orderbookUtil;
