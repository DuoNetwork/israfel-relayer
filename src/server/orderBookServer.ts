import {
	Constants as WrapperConstants,
	DualClassWrapper,
	Web3Wrapper
} from '@finbook/duo-contract-wrapper';
import {
	Constants,
	ILiveOrder,
	IOrderBook,
	IOrderBookLevelUpdate,
	IOrderBookSnapshot,
	IOrderBookSnapshotUpdate,
	OrderBookUtil,
	Util
} from '@finbook/israfel-common';
import { ONE_MINUTE_MS } from '../common/constants';
import { IOption, IOrderQueueItem, IOrderUpdate } from '../common/types';
import dynamoUtil from '../utils/dynamoUtil';
import orderBookPersistenceUtil from '../utils/orderBookPersistenceUtil';
import orderMatchingUtil from '../utils/orderMatchingUtil';
import orderPersistenceUtil from '../utils/orderPersistenceUtil';

class OrderBookServer {
	public pair: string = 'pair';
	public liveOrders: { [orderHash: string]: ILiveOrder } = {};
	public pendingUpdates: IOrderQueueItem[] = [];
	public loadingOrders: boolean = true;
	public orderSnapshotSequence: number = 0;
	public orderBook: IOrderBook = {
		bids: [],
		asks: []
	};
	public orderBookSnapshot: IOrderBookSnapshot = {
		pair: 'pair',
		version: 0,
		bids: [],
		asks: []
	};
	public processedUpdates: { [orderHash: string]: number } = {};
	public custodianInTrading: boolean = false;

	public async handleOrderUpdate(channel: string, orderQueueItem: IOrderQueueItem) {
		Util.logDebug('receive update from channel: ' + channel);

		if (orderQueueItem.requestor === Constants.DB_ORDER_BOOKS) {
			Util.logDebug('ignore order update requested by self');
			return;
		}

		const { method, liveOrder } = orderQueueItem;
		const orderHash = liveOrder.orderHash;

		if (!this.custodianInTrading) {
			Util.logDebug('custodian not in trading, terminate incoming order');
			if (method !== Constants.DB_TERMINATE) this.terminateOrder(orderHash);
			return;
		}

		if (this.loadingOrders) {
			this.pendingUpdates.push(orderQueueItem);
			Util.logDebug('loading orders, queue update');
			return;
		}

		if (![Constants.DB_ADD, Constants.DB_UPDATE, Constants.DB_TERMINATE].includes(method)) {
			Util.logDebug('invalid method, ignore');
			return;
		}

		if (
			liveOrder.currentSequence <= this.orderSnapshotSequence ||
			(this.processedUpdates[orderHash] &&
				this.processedUpdates[orderHash] >= liveOrder.currentSequence)
		) {
			Util.logDebug('obsolete order, ignore');
			return;
		}

		this.processedUpdates[orderHash] = liveOrder.currentSequence;
		if (method === Constants.DB_TERMINATE && !this.liveOrders[orderHash]) {
			Util.logDebug('terminating order not found in cache, ignore');
			return;
		}

		const orderBookLevelUpdates: IOrderBookLevelUpdate[] = [
			this.updateOrderBook({
				liveOrder: orderQueueItem.liveOrder,
				method: method
			})
		];
		const leftLiveOrder = orderQueueItem.liveOrder;
		if (method !== Constants.DB_TERMINATE && leftLiveOrder.balance > 0) {
			const matchinResult = orderMatchingUtil.findMatchingOrders(
				this.orderBook,
				this.liveOrders,
				true
			);
			matchinResult.orderMatchRequests.forEach(orderToMatch =>
				orderMatchingUtil.queueMatchRequest(orderToMatch)
			);
			orderBookLevelUpdates.push(...matchinResult.orderBookLevelUpdates);
		}

		await this.updateOrderBookSnapshot(orderBookLevelUpdates);
	}

	public updateOrderBook(orderUpdate: IOrderUpdate) {
		const liveOrder = orderUpdate.liveOrder;
		const method = orderUpdate.method;
		const orderHash = liveOrder.orderHash;
		const count = OrderBookUtil.updateOrderBook(
			this.orderBook,
			{
				orderHash: orderHash,
				price: liveOrder.price,
				balance: liveOrder.balance,
				initialSequence: liveOrder.initialSequence
			},
			liveOrder.side === Constants.DB_BID,
			method === Constants.DB_TERMINATE
		);

		const orderBookLevelUpdates: IOrderBookLevelUpdate = {
			price: liveOrder.price,
			change:
				(method === Constants.DB_TERMINATE ? 0 : liveOrder.balance) -
				(this.liveOrders[orderHash] ? this.liveOrders[orderHash].balance : 0),
			count: count,
			side: liveOrder.side
		};

		if (method !== Constants.DB_TERMINATE) this.liveOrders[orderHash] = liveOrder;
		else delete this.liveOrders[orderHash];

		return orderBookLevelUpdates;
	}

	public async updateOrderBookSnapshot(orderBookLevelUpdates: IOrderBookLevelUpdate[]) {
		const orderBookSnapshotUpdate: IOrderBookSnapshotUpdate = {
			pair: this.pair,
			updates: [],
			prevVersion: this.orderBookSnapshot.version,
			version: Util.getUTCNowTimestamp()
		};

		for (const update of orderBookLevelUpdates)
			orderBookSnapshotUpdate.updates.push({
				price: update.price,
				change: update.change,
				count: update.count,
				side: update.side
			});

		OrderBookUtil.updateOrderBookSnapshot(this.orderBookSnapshot, orderBookSnapshotUpdate);
		await orderBookPersistenceUtil.publishOrderBookUpdate(
			this.pair,
			this.orderBookSnapshot,
			orderBookSnapshotUpdate
		);
	}

	public updateOrderSequences() {
		this.processedUpdates = {};
		let maxSequence = 0;
		for (const orderHash in this.liveOrders) {
			const liveOrder = this.liveOrders[orderHash];
			maxSequence = Math.max(maxSequence, liveOrder.currentSequence);
			this.processedUpdates[orderHash] = liveOrder.currentSequence;
		}

		this.orderSnapshotSequence = maxSequence;
	}

	public async loadLiveOrders() {
		if (this.custodianInTrading) {
			this.loadingOrders = true;
			this.liveOrders = await orderPersistenceUtil.getAllLiveOrdersInPersistence(this.pair);
			Util.logInfo('loaded live orders : ' + Object.keys(this.liveOrders).length);
			this.updateOrderSequences();
			this.orderBook = OrderBookUtil.constructOrderBook(this.liveOrders);
			Util.logDebug('start matchig ordderBook');
			const matchingResult = orderMatchingUtil.findMatchingOrders(
				this.orderBook,
				this.liveOrders,
				false
			);
			matchingResult.orderMatchRequests.forEach(omr =>
				orderMatchingUtil.queueMatchRequest(omr)
			);
			Util.logInfo('completed matching orderBook as a whole in cold start');
			this.orderBookSnapshot = OrderBookUtil.renderOrderBookSnapshot(
				this.pair,
				this.orderBook
			);
			await orderBookPersistenceUtil.publishOrderBookUpdate(
				this.pair,
				this.orderBookSnapshot
			);
			this.loadingOrders = false;
			for (const updateItem of this.pendingUpdates)
				await this.handleOrderUpdate('pending', updateItem);
			this.pendingUpdates = [];
		}
	}

	public terminateOrder(orderHash: string) {
		return orderPersistenceUtil.persistOrder({
			method: Constants.DB_TERMINATE,
			status: Constants.DB_RESET,
			requestor: Constants.DB_ORDER_BOOKS,
			pair: this.pair,
			orderHash: orderHash
		});
	}

	public async checkCustodianState(dualClassWrapper: DualClassWrapper) {
		const state = await dualClassWrapper.getStates();
		this.custodianInTrading = state.state === WrapperConstants.CTD_TRADING;
		if (!this.custodianInTrading) {
			const prevVersion = this.orderBookSnapshot.version;
			const updates = [
				...this.orderBookSnapshot.bids.map(bid => ({
					price: bid.price,
					change: bid.balance,
					count: -bid.count,
					side: Constants.DB_BID
				})),
				...this.orderBookSnapshot.asks.map(ask => ({
					price: ask.price,
					change: ask.balance,
					count: -ask.count,
					side: Constants.DB_ASK
				}))
			];
			this.orderBook = OrderBookUtil.constructOrderBook({});
			this.orderBookSnapshot = OrderBookUtil.renderOrderBookSnapshot(
				this.pair,
				this.orderBook
			);
			await orderBookPersistenceUtil.publishOrderBookUpdate(
				this.pair,
				this.orderBookSnapshot,
				{
					pair: this.pair,
					updates: updates,
					prevVersion: prevVersion,
					version: this.orderBookSnapshot.version
				}
			);
			for (const orderHash in this.liveOrders) await this.terminateOrder(orderHash);
			this.liveOrders = {};
			this.pendingUpdates = [];
		}
	}

	public async initialize(dualClassWrapper: DualClassWrapper) {
		await this.checkCustodianState(dualClassWrapper);
		global.setInterval(() => this.checkCustodianState(dualClassWrapper), 10000);

		orderPersistenceUtil.subscribeOrderUpdate(this.pair, (channel, orderQueueItem) =>
			this.handleOrderUpdate(channel, orderQueueItem)
		);

		await this.loadLiveOrders();
		global.setInterval(() => this.loadLiveOrders(), ONE_MINUTE_MS * 15);
	}

	public async startServer(option: IOption) {
		this.pair = option.token + '|' + Constants.TOKEN_WETH;
		const tokens = await dynamoUtil.scanTokens();
		const token = tokens.find(t => t.code === option.token);
		if (!token) {
			Util.logInfo('Invalid token, exit');
			return;
		}

		let infura = { token: '' };
		try {
			infura = require('../keys/infura.json');
		} catch (err) {
			Util.logError(JSON.stringify(err));
		}

		this.initialize(
			new DualClassWrapper(
				new Web3Wrapper(
					null,
					'infura',
					(option.env === Constants.DB_LIVE
						? Constants.PROVIDER_INFURA_MAIN
						: Constants.PROVIDER_INFURA_KOVAN) +
						'/' +
						infura.token,
					option.env === Constants.DB_LIVE
				),
				token.custodian
			)
		);
		if (option.server) {
			dynamoUtil.updateStatus(this.pair);
			global.setInterval(
				() => dynamoUtil.updateStatus(this.pair, Object.keys(this.liveOrders).length),
				15000
			);
		}
	}
}

const orderBookServer = new OrderBookServer();
export default orderBookServer;
