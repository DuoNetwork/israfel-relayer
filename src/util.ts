import { BigNumber } from '0x.js';
import moment from 'moment';
import * as os from 'os';
import * as CST from './constants';
import { IOption } from './types';

class Util {
	public logLevel: string = CST.LOG_INFO;

	public logInfo(text: any): void {
		this.log(text, CST.LOG_INFO);
	}

	public logDebug(text: any): void {
		this.log(text, CST.LOG_DEBUG);
	}

	public logError(text: any): void {
		this.log(text, CST.LOG_ERROR);
	}

	private log(text: any, level: string): void {
		if (CST.LOG_RANKING[this.logLevel] >= CST.LOG_RANKING[level])
			console.log(`${moment().format('HH:mm:ss.SSS')} [${level}]: ` + text);
	}

	public isNumber(input: any): boolean {
		const num = Number(input);
		return isFinite(num) && !isNaN(num);
	}

	public isEmptyObject(obj: object | undefined | null): boolean {
		if (!obj) return true;

		for (const prop in obj) if (obj.hasOwnProperty(prop)) return false;

		return true;
	}

	public getHostName() {
		return os.hostname();
	}

	public getRandomFutureDateInSeconds() {
		return new BigNumber(Date.now() + CST.TEN_MINUTES_MS).div(CST.ONE_SECOND_MS).ceil();
	}

	public stringToBN(value: string): BigNumber {
		return new BigNumber(value);
	}

	public defaultOption: IOption = {
		live: false,
		token: 'ZRX',
		amount: 1,
		maker: 0,
		spender: 1
	};

	public getUTCNowTimestamp() {
		return moment().valueOf();
	}

	public parseOptions(argv: string[]): IOption {
		const option: IOption = this.defaultOption;

		for (let i = 3; i < argv.length; i++) {
			option.live = process.argv.includes('live');
			const args = argv[i].split('=');
			switch (args[0]) {
				case 'token':
					option.token = args[1] || option.token;
					break;
				case 'amount':
					option.amount = Number(args[1]) || option.amount;
					break;
				case 'maker':
					option.maker = Number(args[1]) || option.maker;
					break;
				case 'spender':
					option.maker = Number(args[1]) || option.spender;
					break;
				default:
					break;
			}
		}

		return option;
	}
}

const util = new Util();
export default util;
