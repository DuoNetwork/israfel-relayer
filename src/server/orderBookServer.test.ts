import * as CST from '../common/constants';
// import { ILiveOrder } from '../common/types';
import liveOrders from '../samples/test/liveOrders.json';
import orderBookUtil from '../utils/orderBookUtil';
import orderPersistenceUtil from '../utils/orderPersistenceUtil';
import util from '../utils/util';
import orderBookServer from './orderBookServer';

// const liveOrders: { [orderhash: string]: ILiveOrder } =
orderBookServer.pair = 'pair';
const channel = 'xxx|xxx|pair';

const orderQueueItem = {
	method: CST.DB_ADD,
	liveOrder: liveOrders['orderHash1']
};
test('handleOrderUpdate, pair wrong', async () => {
	util.getUTCNowTimestamp = jest.fn(() => 123456789);
	orderBookUtil.publishOrderBookUpdate = jest.fn(() => Promise.resolve(true));
	await orderBookServer.handleOrderUpdate('xxx', orderQueueItem);
	expect(orderBookServer.pendingUpdates.length).toBe(0);
	expect(orderBookServer.processedUpdates).toMatchObject({});
	expect(util.getUTCNowTimestamp as jest.Mock).not.toBeCalled();
	expect(orderBookUtil.publishOrderBookUpdate as jest.Mock).not.toBeCalled();
});

test('handleOrderUpdate, method wrong, not loadingPairs', async () => {
	orderQueueItem.method = 'xxx';
	orderBookServer.loadingOrders = false;
	util.getUTCNowTimestamp = jest.fn(() => 123456789);
	orderBookUtil.publishOrderBookUpdate = jest.fn(() => Promise.resolve(true));
	await orderBookServer.handleOrderUpdate(channel, orderQueueItem);
	expect(orderBookServer.pendingUpdates.length).toBe(0);
	expect(orderBookServer.processedUpdates).toMatchObject({});
	expect(util.getUTCNowTimestamp as jest.Mock).not.toBeCalled();
	expect(orderBookUtil.publishOrderBookUpdate as jest.Mock).not.toBeCalled();
});

test('handleOrderUpdate,loading Pairs', async () => {
	orderBookServer.loadingOrders = true;
	orderQueueItem.method = CST.DB_ADD;
	util.getUTCNowTimestamp = jest.fn(() => 123456789);
	orderBookUtil.publishOrderBookUpdate = jest.fn(() => Promise.resolve(true));
	await orderBookServer.handleOrderUpdate(channel, orderQueueItem);
	expect(orderBookServer.pendingUpdates).toMatchSnapshot();
	expect(orderBookServer.pendingUpdates.length).toBe(1);
	expect(orderBookServer.processedUpdates).toMatchObject({});
	expect(util.getUTCNowTimestamp as jest.Mock).not.toBeCalled();
	expect(orderBookUtil.publishOrderBookUpdate as jest.Mock).not.toBeCalled();
});

test('updateOrderSequences', async () => {
	orderBookServer.liveOrders = liveOrders;
	orderBookServer.updateOrderSequences();
	expect(orderBookServer.orderSnapshotSequence).toMatchSnapshot();
});

test('loadLiveOrders', async () => {
	orderPersistenceUtil.getAllLiveOrdersInPersistence = jest.fn(() => Promise.resolve(liveOrders));
	orderBookUtil.publishOrderBookUpdate = jest.fn(() => Promise.resolve(true));
	orderBookServer.handleOrderUpdate = jest.fn(() => Promise.resolve());
	expect(orderBookServer.loadingOrders).toBeTruthy();
	await orderBookServer.loadLiveOrders();
	expect(orderBookServer.orderBook).toMatchSnapshot();
	expect(orderBookServer.orderBookSnapshot).toMatchSnapshot();
	expect((orderBookUtil.publishOrderBookUpdate as jest.Mock).mock.calls).toMatchSnapshot();
	expect((orderBookServer.handleOrderUpdate as jest.Mock).mock.calls).toMatchSnapshot();
	expect(orderBookServer.pendingUpdates).toMatchSnapshot();
	expect(orderBookServer.loadingOrders).toBeFalsy();
});
