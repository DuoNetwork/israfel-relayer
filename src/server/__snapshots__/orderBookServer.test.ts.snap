// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`handleOrderUpdate, current sequence too small 1`] = `Object {}`;

exports[`handleOrderUpdate, method wrong, not loadingPairs 1`] = `Object {}`;

exports[`handleOrderUpdate, order processed already 1`] = `
Object {
  "orderHash1": 200,
}
`;

exports[`handleOrderUpdate, terminate a non existing liveOrder 1`] = `
Object {
  "orderHash1": 5,
}
`;

exports[`handleOrderUpdate, update orderBooks , add 1`] = `
Object {
  "orderHash1": 5,
}
`;

exports[`handleOrderUpdate, update orderBooks , add 2`] = `
Object {
  "orderHash1": Object {
    "account": "account1",
    "amount": 10,
    "balance": 5,
    "createdAt": 1234567890,
    "currentSequence": 5,
    "expiry": 1234567890000,
    "fee": 1,
    "feeAsset": "feeAsset",
    "fill": 1,
    "initialSequence": 1,
    "orderHash": "orderHash1",
    "pair": "pair",
    "price": 100,
    "side": "bid",
  },
}
`;

exports[`handleOrderUpdate, update orderBooks , add 3`] = `
Array [
  Array [
    "code1|code2",
    Object {
      "asks": Array [],
      "bids": Array [
        Object {
          "balance": 5,
          "count": 1,
          "price": 100,
        },
      ],
      "pair": "pair",
      "version": 123456789,
    },
    Object {
      "pair": "code1|code2",
      "prevVersion": 0,
      "updates": Array [
        Object {
          "change": 5,
          "count": 1,
          "price": 100,
          "side": "bid",
        },
      ],
      "version": 123456789,
    },
  ],
]
`;

exports[`handleOrderUpdate, update orderBooks , add 4`] = `
Object {
  "asks": Array [],
  "bids": Array [
    Object {
      "balance": 5,
      "initialSequence": 1,
      "orderHash": "orderHash1",
      "price": 100,
    },
  ],
}
`;

exports[`handleOrderUpdate, update orderBooks , add 5`] = `
Object {
  "asks": Array [],
  "bids": Array [
    Object {
      "balance": 5,
      "count": 1,
      "price": 100,
    },
  ],
  "pair": "pair",
  "version": 123456789,
}
`;

exports[`handleOrderUpdate, update orderBooks , terminate 1`] = `
Object {
  "orderHash1": 5,
}
`;

exports[`handleOrderUpdate, update orderBooks , terminate 2`] = `Object {}`;

exports[`handleOrderUpdate, update orderBooks , terminate 3`] = `
Array [
  Array [
    "code1|code2",
    Object {
      "asks": Array [],
      "bids": Array [],
      "pair": "pair",
      "version": 123456789,
    },
    Object {
      "pair": "code1|code2",
      "prevVersion": 123456789,
      "updates": Array [
        Object {
          "change": -5,
          "count": -1,
          "price": 100,
          "side": "bid",
        },
      ],
      "version": 123456789,
    },
  ],
]
`;

exports[`handleOrderUpdate, update orderBooks , terminate 4`] = `
Object {
  "asks": Array [],
  "bids": Array [],
}
`;

exports[`handleOrderUpdate, update orderBooks , terminate 5`] = `
Object {
  "asks": Array [],
  "bids": Array [],
  "pair": "pair",
  "version": 123456789,
}
`;

exports[`handleOrderUpdate,loading Pairs 1`] = `
Array [
  Object {
    "liveOrder": Object {
      "account": "account1",
      "amount": 10,
      "balance": 5,
      "createdAt": 1234567890,
      "currentSequence": 5,
      "expiry": 1234567890000,
      "fee": 1,
      "feeAsset": "feeAsset",
      "fill": 1,
      "initialSequence": 1,
      "orderHash": "orderHash1",
      "pair": "pair",
      "price": 100,
      "side": "bid",
    },
    "method": "add",
    "requestor": "requestor",
    "status": "status",
  },
]
`;

exports[`handleOrderUpdate,loading Pairs 2`] = `
Object {
  "orderHash1": 5,
}
`;

exports[`loadLiveOrders 1`] = `"orderBook"`;

exports[`loadLiveOrders 2`] = `"orderBookSnapshot"`;

exports[`loadLiveOrders 3`] = `
Array [
  Array [
    "liveOrders",
  ],
]
`;

exports[`loadLiveOrders 4`] = `
Array [
  Array [
    "code1|code2",
    "orderBook",
  ],
]
`;

exports[`loadLiveOrders 5`] = `
Array [
  Array [
    "orderBook",
    "liveOrders",
  ],
]
`;

exports[`loadLiveOrders 6`] = `
Array [
  Array [
    null,
    "code1|code2",
    Array [
      "ordersToMatch",
    ],
  ],
]
`;

exports[`loadLiveOrders 7`] = `
Array [
  Array [
    "code1|code2",
    "orderBookSnapshot",
  ],
]
`;

exports[`loadLiveOrders 8`] = `
Array [
  Array [
    "pending",
    Object {
      "liveOrder": Object {
        "account": "account1",
        "amount": 10,
        "balance": 5,
        "createdAt": 1234567890,
        "currentSequence": 5,
        "expiry": 1234567890000,
        "fee": 1,
        "feeAsset": "feeAsset",
        "fill": 1,
        "initialSequence": 1,
        "orderHash": "orderHash1",
        "pair": "pair",
        "price": 100,
        "side": "bid",
      },
      "method": "add",
      "requestor": "requestor",
      "status": "status",
    },
  ],
]
`;

exports[`updateOrderSequences 1`] = `16`;

exports[`updateOrderSequences 2`] = `
Object {
  "orderHash1": 5,
  "orderHash2": 6,
  "orderHash3": 7,
  "orderHash4": 8,
  "orderHash5": 13,
  "orderHash6": 14,
  "orderHash7": 15,
  "orderHash8": 16,
}
`;
